frlearn/uncategorised/fuzzy_quantifiers

```diff
@@ -0,0 +1,413 @@
+        mu: function
+            Define a monotone measure on set_v
+        """
+    set_sorted = sorted(set_v, key=lambda x: f(x))
+    subsets = []
+    integral = 0
+    for i in range(len(set_sorted)):
+        subsets.append([])
+        subsets[i] = set_sorted[i:]
+        if i == 0:
+            integral = mu(subsets[0]) * f(set_sorted[0])
+        else:
+            integral += mu(subsets[i]) * (f(set_sorted[i]) - f(set_sorted[i - 1]))
+    return integral
+
+
+def more_than(k, p):
+    """"
+    Examples of RIM quantifiers which represent the quantifiers “more than 100 ∗ 𝑘%”
+    parameters:
+    ----------
+    k, p: float
+    ”"""
+    if p > k:
+        return 1
+    else:
+        return 0
+
+
+def exists(p):
+    return more_than(0, p)
+
+
+def at_least(k, p):
+    """"
+        Examples of RIM quantifiers which represent the quantifiers “at least 100 ∗ 𝑘%”
+        parameters:
+        ----------
+        k, p: float
+        ”"""
+    if p >= k:
+        return 1
+    else:
+        return 0
+
+
+def for_all(p):
+    return at_least(1, p)
+
+
+def zadeh_function(a, b, p):
+    """"
+    Zadeh’s S-function
+    parameters: 
+    ----------
+    a, b, p: float such that 0 <= a<b <=1
+    """
+    if p <= a:
+        return 0
+    elif a <= p <= (a + b) / 2:
+        return (2 * (p - a) ** 2) / (b - a) ** 2
+    elif (a + b) / 2 <= p <= b:
+        return 1 - ((2 * (p - b) ** 2) / (b - a) ** 2)
+    else:
+        return 1
+
+
+def most(p):
+    return zadeh_function(0.3, 0.9, p)
+
+
+def some(p):
+    return zadeh_function(0.1, 0.4, p)
+
+
+def z_eval_unary_sentence(set_v, rim, A):
+    """
+    Evaluation of unary sentences of the form “Λ X’s are A’s” in Zadeh’s model, where Λ is a RIM quantifier
+    parameters:
+    ----------
+    rim: function
+        Definition of the RIM quantifier
+    A: table (list)
+       List of value of the Fuzzy set
+    set_v: table (list)
+           The base set X on which the fuzzy set is define
+    """
+
+    return rim(sum(A) / len(set_v))
+
+
+def z_eval_binary_sentence(rim, A, B):
+    """
+    Evaluation of binary sentences of the form “Λ A’s are B’s” in Zadeh’s model, where Λ is a RIM quantifier
+    parameters:
+    ----------
+    rim: function
+         Definition of the RIM quantifier
+    A, B: table (list)
+          List of value of the Fuzzy sets
+    """
+    return rim(sum([el for el in A if el in B]) / sum(A))
+
+
+def y_eval_unary_sentence(set_v, rim, A):
+    """
+        Evaluation of unary sentences of the form “Λ X’s are A’s” in Yager’s OWA model, where Λ is a RIM quantifier
+        parameters:
+        ----------
+        rim: function
+            Definition of the RIM quantifier
+        A: function
+           Function defining the fuzzy set A
+        set_v: table (list)
+               The base set X on which the fuzzy set is defined
+        """
+    def mu(x):
+        return rim((len(x)/len(set_v)))
+    return choquet_integral(set_v, A, mu)
```

Comment:
This is just an OWA operator, so defining a measure and then using the general choquet_integral function is inefficient. Same thing for other functions, for specific measures you can simplify the choquet integral.
Also remove set_v and define A as a list.

Reply:
I've just implemented the definitions of quantificators present in the parper 'Fuzzy rough sets based on fuzzy quantification, Adnan Theerens ∗, Chris Cornelis'. As for the efficiency of these quantificators, I haven't thought about it.